### ISSUE TITLE
BGZF compression test failures

### ISSUE THREAD

**Issue by sagitter** [biopython/biopython#4553]
Hi all.

### Setup

I am reporting a problem with Biopython version, Python version, and operating
system as follows:

Fedora 40 x86_64
Python-3.12.1
biopython-1.82

### Expected behaviour

No errors

### Actual behaviour
```
Check get expected TypeError from BgzfBlocks. ... ok
test_append_mode (test_bgzf.BgzfTests.test_append_mode) ... ok
test_bam_ex1 (test_bgzf.BgzfTests.test_bam_ex1)
Reproduce BGZF compression for BAM file. ... FAIL
test_double_flush (test_bgzf.BgzfTests.test_double_flush) ... ok
test_example_cor6 (test_bgzf.BgzfTests.test_example_cor6)
Reproduce BGZF compression for cor6_6.gb GenBank file. ... FAIL
test_example_fastq (test_bgzf.BgzfTests.test_example_fastq)
Reproduce BGZF compression for a FASTQ file. ... FAIL
test_example_gb (test_bgzf.BgzfTests.test_example_gb)
Reproduce BGZF compression for NC_000932 GenBank file. ... FAIL
test_example_wnts_xml (test_bgzf.BgzfTests.test_example_wnts_xml)
Reproduce BGZF compression for wnts.xml BLAST file. ... FAIL
test_iter_bam_ex1 (test_bgzf.BgzfTests.test_iter_bam_ex1)
Check iteration over SamBam/ex1.bam. ... ok
...
======================================================================
FAIL: test_bam_ex1 (test_bgzf.BgzfTests.test_bam_ex1)
Reproduce BGZF compression for BAM file.
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/builddir/build/BUILD/biopython-1.82/Tests/test_bgzf.py", line 305, in test_bam_ex1
    self.check_blocks("SamBam/ex1.bam", temp_file)
  File "/builddir/build/BUILD/biopython-1.82/Tests/test_bgzf.py", line 58, in check_blocks
    self.assertEqual(old, new)
AssertionError: Lists differ: [(0, 18239, 0, 65536), (18239, 18223, 65536, 655[180 chars], 0)] != [(0, 18078, 0, 65536), (18078, 18147, 65536, 655[180 chars], 0)]

First differing element 0:
(0, 18239, 0, 65536)
(0, 18078, 0, 65536)

Diff is 780 characters long. Set self.maxDiff to None to see it.

======================================================================
FAIL: test_example_cor6 (test_bgzf.BgzfTests.test_example_cor6)
Reproduce BGZF compression for cor6_6.gb GenBank file.
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/builddir/build/BUILD/biopython-1.82/Tests/test_bgzf.py", line 327, in test_example_cor6
    self.check_blocks("GenBank/cor6_6.gb.bgz", temp_file)
  File "/builddir/build/BUILD/biopython-1.82/Tests/test_bgzf.py", line 58, in check_blocks
    self.assertEqual(old, new)
AssertionError: Lists differ: [(0, 4350, 0, 14967), (4350, 28, 14967, 0)] != [(0, 4345, 0, 14967), (4345, 28, 14967, 0)]

First differing element 0:
(0, 4350, 0, 14967)
(0, 4345, 0, 14967)

- [(0, 4350, 0, 14967), (4350, 28, 14967, 0)]
?           ---             -

+ [(0, 4345, 0, 14967), (4345, 28, 14967, 0)]
?        + ++              +


======================================================================
FAIL: test_example_fastq (test_bgzf.BgzfTests.test_example_fastq)
Reproduce BGZF compression for a FASTQ file.
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/builddir/build/BUILD/biopython-1.82/Tests/test_bgzf.py", line 315, in test_example_fastq
    self.check_blocks("Quality/example.fastq.bgz", temp_file)
  File "/builddir/build/BUILD/biopython-1.82/Tests/test_bgzf.py", line 58, in check_blocks
    self.assertEqual(old, new)
AssertionError: Lists differ: [(0, 147, 0, 234), (147, 28, 234, 0)] != [(0, 146, 0, 234), (146, 28, 234, 0)]

First differing element 0:
(0, 147, 0, 234)
(0, 146, 0, 234)

- [(0, 147, 0, 234), (147, 28, 234, 0)]
?        ^              ^

+ [(0, 146, 0, 234), (146, 28, 234, 0)]
?        ^              ^


======================================================================
FAIL: test_example_gb (test_bgzf.BgzfTests.test_example_gb)
Reproduce BGZF compression for NC_000932 GenBank file.
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/builddir/build/BUILD/biopython-1.82/Tests/test_bgzf.py", line 321, in test_example_gb
    self.check_blocks("GenBank/NC_000932.gb.bgz", temp_file)
  File "/builddir/build/BUILD/biopython-1.82/Tests/test_bgzf.py", line 58, in check_blocks
    self.assertEqual(old, new)
AssertionError: Lists differ: [(0, 15073, 0, 65536), (15073, 17857, 65536, 65[117 chars], 0)] != [(0, 14707, 0, 65536), (14707, 17694, 65536, 65[117 chars], 0)]

First differing element 0:
(0, 15073, 0, 65536)
(0, 14707, 0, 65536)

- [(0, 15073, 0, 65536),
?       ^  -

+ [(0, 14707, 0, 65536),
?       ^^

-  (15073, 17857, 65536, 65536),
?    ^  -    ^^^

+  (14707, 17694, 65536, 65536),
?    ^^      ^^^

-  (32930, 22144, 131072, 65536),
?     ^^   -  ^^

+  (32401, 21975, 131072, 65536),
?     ^ +    ^^^

-  (55074, 22230, 196608, 65536),
?    ^^ ^    --

+  (54376, 22066, 196608, 65536),
?    ^^ ^     ++

-  (77304, 14939, 262144, 43478),
?    ^^^     ^^

+  (76442, 14859, 262144, 43478),
?    ^ ++    ^^

-  (92243, 28, 305622, 0)]
?    ^^^

+  (91301, 28, 305622, 0)]
?    ^ ++


======================================================================
FAIL: test_example_wnts_xml (test_bgzf.BgzfTests.test_example_wnts_xml)
Reproduce BGZF compression for wnts.xml BLAST file.
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/builddir/build/BUILD/biopython-1.82/Tests/test_bgzf.py", line 333, in test_example_wnts_xml
    self.check_blocks("Blast/wnts.xml.bgz", temp_file)
  File "/builddir/build/BUILD/biopython-1.82/Tests/test_bgzf.py", line 58, in check_blocks
    self.assertEqual(old, new)
AssertionError: Lists differ: [(0, 5280, 0, 65536), (5280, 5477, 65536, 65536)[170 chars], 0)] != [(0, 5209, 0, 65536), (5209, 5406, 65536, 65536)[170 chars], 0)]

First differing element 0:
(0, 5280, 0, 65536)
(0, 5209, 0, 65536)

Diff is 746 characters long. Set self.maxDiff to None to see it.
```

### Steps to reproduce

Running 

+ pushd Tests
~/build/BUILD/biopython-1.82/Tests ~/build/BUILD/biopython-1.82
+ PATH=/builddir/build/BUILDROOT/python-biopython-1.82-1.fc40.x86_64/usr/bin:/usr/bin:/bin:/usr/sbin:/sbin:/usr/local/sbin
+ PYTHONPATH=/builddir/build/BUILDROOT/python-biopython-1.82-1.fc40.x86_64/usr/lib64/python3.12/site-packages:/builddir/build/BUILDROOT/python-biopython-1.82-1.fc40.x86_64/usr/lib/python3.12/site-packages
+ PYTHONDONTWRITEBYTECODE=1
+ PYTEST_ADDOPTS=' --ignore=/builddir/build/BUILD/biopython-1.82/.pyproject-builddir'
+ PYTEST_XDIST_AUTO_NUM_WORKERS=8
+ /usr/bin/python3 run_tests.py --offline --verbose

**Comment by peterjc** [comment]
Hmm. All those failures are from ``test_bgzf.py`` which you can run directly, e.g.

```
$ cd Tests
$ python test_bgzf.py
```

And all from the ``check_blocks`` method, which confirms the block structure of a recently recreated file matches the older provided file.

Can we first check your test files match (and this isn't something strange with however you downloaded the test files):

```
$ md5sum */*.bgz */*.bam
a09a73135a3a6a4be17a3cbe2cbb5a7c Blast/wnts.xml.bgz
dab7defcbb7fbc946cf2779b9df8036b GenBank/NC_000932.gb.bgz
8e4be547729de4c7ccd9e54c51784abe GenBank/cor6_6.gb.bgz
79289b51cf9c12f652738772899186d5 Quality/example.fastq.bgz
8e3060db899fcfc0e72674c0a56472e1 Quality/example_dos.fastq.bgz
89930f9d337b0120219ce9d12f961b03 SamBam/bam1.bam
ba2d9f15a9b2293fe6923655def5b8e8 SamBam/bam1_sorted.bam
80df494bdfb55e1a09ca482d59eace2d SamBam/bam2.bam
7b8859fe74c4d2d06338322f01c94a80 SamBam/bam2_sorted.bam
04dbe28124120274f0dfd19941a180f3 SamBam/ex1.bam
e1af8e584e20f4ff7b4cc1e49921127e SamBam/ex1_header.bam
69fff5f126089205e168ccbb10906570 SamBam/ex1_refresh.bam
```

My guess however is a difference in your GZIP C library, perhaps it is defaulting to a different compression level or something like that...

**Comment by peterjc** [comment]
I *suspect* this is linked to https://pagure.io/fesco/issue/3109 if Fedora 40 has indeed replaced the zlib library with zlib-ng (and this altered what ``import zlib`` does from Python).

**Comment by sagitter** [comment]
> Can we first check your test files match (and this isn't something strange with however you downloaded the test files):
```
a09a73135a3a6a4be17a3cbe2cbb5a7c Blast/wnts.xml.bgz
dab7defcbb7fbc946cf2779b9df8036b GenBank/NC_000932.gb.bgz
8e4be547729de4c7ccd9e54c51784abe GenBank/cor6_6.gb.bgz
79289b51cf9c12f652738772899186d5 Quality/example.fastq.bgz
8e3060db899fcfc0e72674c0a56472e1 Quality/example_dos.fastq.bgz
89930f9d337b0120219ce9d12f961b03 SamBam/bam1.bam
ba2d9f15a9b2293fe6923655def5b8e8 SamBam/bam1_sorted.bam
80df494bdfb55e1a09ca482d59eace2d SamBam/bam2.bam
7b8859fe74c4d2d06338322f01c94a80 SamBam/bam2_sorted.bam
04dbe28124120274f0dfd19941a180f3 SamBam/ex1.bam
e1af8e584e20f4ff7b4cc1e49921127e SamBam/ex1_header.bam
69fff5f126089205e168ccbb10906570 SamBam/ex1_refresh.bam

```

```
sagitter@localhost:/var/lib/mock/fedora-rawhide-x86_64/root/builddir/build/BUILD/biopython-1.82/Tests$ md5sum Blast/wnts.xml.bgz GenBank/NC_000932.gb.bgz GenBank/cor6_6.gb.bgz Quality/example.fastq.bgz Quality/example_dos.fastq.bgz SamBam/*.bam
a09a73135a3a6a4be17a3cbe2cbb5a7c  Blast/wnts.xml.bgz
dab7defcbb7fbc946cf2779b9df8036b  GenBank/NC_000932.gb.bgz
8e4be547729de4c7ccd9e54c51784abe  GenBank/cor6_6.gb.bgz
79289b51cf9c12f652738772899186d5  Quality/example.fastq.bgz
8e3060db899fcfc0e72674c0a56472e1  Quality/example_dos.fastq.bgz
89930f9d337b0120219ce9d12f961b03  SamBam/bam1.bam
ba2d9f15a9b2293fe6923655def5b8e8  SamBam/bam1_sorted.bam
80df494bdfb55e1a09ca482d59eace2d  SamBam/bam2.bam
7b8859fe74c4d2d06338322f01c94a80  SamBam/bam2_sorted.bam
04dbe28124120274f0dfd19941a180f3  SamBam/ex1.bam
e1af8e584e20f4ff7b4cc1e49921127e  SamBam/ex1_header.bam
69fff5f126089205e168ccbb10906570  SamBam/ex1_refresh.bam
```

**Comment by peterjc** [comment]
By eye those checksums match up. Do you know anything about the Fedora 40
zlib-ng changes and if that would explain this?

**Comment by sagitter** [comment]
>Do you know anything about the Fedora 40 zlib-ng changes and if that would explain this?

No more of what it was reported in the [Change](https://fedoraproject.org/wiki/Changes/ZlibNGTransition#Benefit_to_Fedora). Sorry.

**Comment by musicinmybrain** [comment]
> By eye those checksums match up. Do you know anything about the Fedora 40 zlib-ng changes and if that would explain this?

Well, based on a quick examination of [`BgzfBlocks`](https://github.com/biopython/biopython/blob/2204ad0b0f92c4b8ed769ce4c1a8a3987b048c6f/Bio/bgzf.py#L348), this test is comparing the offsets and lengths of the compressed data blocks, so it’s expecting a bit-identical compressed data stream. That’s not even guaranteed across versions of classic `zlib`, but it’s been the case in practice since there hasn’t been much change to the compression strategies over time.

What `zlib-ng` guarantees is API- and ABI-compatibility with `zlib` when used via the `zlib-compat` library, and that the compressed streams are interoperable (it can read any stream `zlib` wrote, and `zlib` can read any stream it writes). It doesn’t guarantee byte-for-byte identical compressed output at a particular compression level, because that would be difficult to ensure and far too constraining on the implementation.

Perhaps you can compare data starts and lengths only, ignoring the compressed starts and lengths, e.g instead of

https://github.com/biopython/biopython/blob/2204ad0b0f92c4b8ed769ce4c1a8a3987b048c6f/Tests/test_bgzf.py#L58

something like:

```python
    old_data = [(ds, dl) for _, _, ds, dl in old]
    new_data = [(ds, dl) for _, _, ds, dl in new]
    self.assertEqual(old_data, new_data)
```

**Comment by peterjc** [comment]
Yes - what I am only guessing is Python on Fedora 40 gets zlib-ng behavior.
And if so can we detect this at runtime (and skip that test)?

**Comment by musicinmybrain** [comment]
> Yes - what I am only guessing is Python on Fedora 40 gets zlib-ng behavior. And if so can we detect this at runtime (and skip that test)?

I mean, it’s probably *possible*, but are you sure you *need* this test to depend on exact reproducibility of the compressed stream? A file that encodes the same data, but where the encoder made slightly different choices, is no less correct…

**Comment by peterjc** [comment]
Exactly - and as you noted, this zlib behaviour has been stable for many years, so it was pragmatic to test to identical compression. Right now zlib-ng is niche, but may grow in popularity. I'd prefer to skip this test under zlib-ng, but removing the test is my fall back position.

**Comment by ghuls** [comment]
Preferably the BGZF compression code would use [deflate](https://github.com/dcwatson/deflate) (if installed) for compression/decompression as HTSlib recommends libdeflate too.

**Comment by peterjc** [comment]
The Biopython BGZF code could in principle use libdeflate, but that would be an extra external dependency so I am reluctant to do so. Perhaps it wouldn't be too painful to use it if present and fall back on the standard libary as now, but it complicates our test matrix.

Dropping this test would probably make sense under that model too, and there is no reason to expect libdeflate to give exactly the same compressed bits either.

**Comment by musicinmybrain** [comment]
Tested in F40 and in F41/Rawhide as follows:

- Patch [biopython 1.83 in Fedora](https://src.fedoraproject.org/rpms/python-biopython) to [fix tests failing with expat 2.6.0](https://src.fedoraproject.org/rpms/python-biopython/pull-request/7). Do a successful build (with `test_bgzf.py` skipped).
- Stop skipping `test_bgzf.py`, and confirm that I can reproduce the test failures as reported in this issue.
- [Apply](https://src.fedoraproject.org/rpms/python-biopython/pull-request/8) https://github.com/biopython/biopython/pull/4696.patch, and confirm the tests pass.

So… it looks good to me.

**Comment by peterjc** [comment]
Thank you! Both those issues should be fixed in our next release - which I hope to do as soon as the numpy2 work is done.
